File: LeetCode/cpp/LC0051_NQueens/NQueens.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
	vector<vector<string>> res;
	
	bool isValid(int row, int col, vector<string> &board, int n) {
		// Check current column
		for (int i = 0; i < row; i++) {
			if (board[i][col] == 'Q') return false;
		}
		
		// Check diag (init_row = row - 1; downward)
		for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
			if (board[i][j] == 'Q') return false;
		}
		
		// Check counter-diag (init_row = row - 1; upward)
		for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
			if (board[i][j] == 'Q') return false;
		}
		
		return true;
	}
	
	void dfs(vector<string> &board, int n, int row) {
		if (row == n) {
			res.push_back(board);
			return;
		}
		
		for (int i = 0; i < n; i++) {
			if (isValid(row, i, board, n)) {
				board[row][i] = 'Q';
				dfs(board, n, row + 1);
				board[row][i] = '.';
			}
		}
	}
	
	vector<vector<string>> solveNQueens(int n) {
		vector<string> board(n, string(n, '.'));
		dfs(board, n, 0);
		return res;
	}
};

File: LeetCode/java/LC0051_N_Queens/NQueens.java
package LC0051_N_Queens;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * <a href="https://leetcode.com/problems/n-queens/">N-Queens</a>
 * 数组；回溯
 */
public class NQueens {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.solveNQueens(4));
    }
}

class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<>();
        char[][] board = new char[n][n];
        for (char[] row : board) {
            Arrays.fill(row, '.');
        }

        dfs(res, board, n, 0);
        return res;
    }

    private void dfs(List<List<String>> res, char[][] board, int n, int row) {
        if (row == n) {
            res.add(construct(board));
            return;
        }

        for (int col = 0; col < n; col++) {
            if (isValid(board, n, row, col)) {
                board[row][col] = 'Q';
                dfs(res, board, n, row + 1);
                board[row][col] = '.';
            }
        }
    }

    private List<String> construct(char[][] board) {
        List<String> res = new ArrayList<>();
        for (char[] row : board) {
            res.add(String.valueOf(row));
        }
        return res;
    }

    private boolean isValid(char[][] board, int n, int row, int col) {
        // Check current column
        for (int i = 0; i < row; i++) {
            if (board[i][col] == 'Q') {
                return false;
            }
        }

        // Check upper left diagonal
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        // Check upper right diagonal
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        return true;
    }
}